File: agents/agent-go/socks.go
package agent

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"sync"
	"sync/atomic"
)

var (
	socksListener net.Listener
	socksMu       sync.Mutex
	socksRunning  atomic.Bool
	socksUser     string
	socksPass     string
)

func startSOCKS5(port int, user, pass string) (string, error) {
	socksMu.Lock()
	defer socksMu.Unlock()

	if socksRunning.Load() {
		stopSOCKS5Locked()
	}

	ln, err := net.Listen("tcp", fmt.Sprintf("0.0.0.0:%d", port))
	if err != nil {
		return "", err
	}
	socksListener = ln
	socksUser = user
	socksPass = pass
	socksRunning.Store(true)

	go func() {
		defer socksRunning.Store(false)
		for {
			conn, err := ln.Accept()
			if err != nil {
				return
			}
			go handleSOCKS5(conn, user, pass)
		}
	}()

	return ln.Addr().String(), nil
}

func stopSOCKS5() {
	socksMu.Lock()
	defer socksMu.Unlock()
	stopSOCKS5Locked()
}

func stopSOCKS5Locked() {
	if socksListener != nil {
		socksListener.Close()
		socksListener = nil
	}
	socksRunning.Store(false)
}

// handleSOCKS5 implements SOCKS5 (RFC 1928) with optional RFC 1929 user/pass auth.
func handleSOCKS5(client net.Conn, user, pass string) {
	defer client.Close()

	// Auth negotiation
	hdr := make([]byte, 2)
	if _, err := io.ReadFull(client, hdr); err != nil {
		return
	}
	if hdr[0] != 0x05 {
		return
	}
	nMethods := int(hdr[1])
	methods := make([]byte, nMethods)
	if _, err := io.ReadFull(client, methods); err != nil {
		return
	}

	if user != "" {
		// Require method 0x02 (user/pass).
		offered := false
		for _, m := range methods {
			if m == 0x02 {
				offered = true
				break
			}
		}
		if !offered {
			client.Write([]byte{0x05, 0xFF}) // no acceptable method
			return
		}
		client.Write([]byte{0x05, 0x02})

		// RFC 1929 sub-negotiation: VER(1) ULEN(1) USER(N) PLEN(1) PASS(M)
		sub := make([]byte, 2)
		if _, err := io.ReadFull(client, sub); err != nil {
			return
		}
		if sub[0] != 0x01 {
			return
		}
		uBuf := make([]byte, sub[1])
		if _, err := io.ReadFull(client, uBuf); err != nil {
			return
		}
		pLen := make([]byte, 1)
		if _, err := io.ReadFull(client, pLen); err != nil {
			return
		}
		pBuf := make([]byte, pLen[0])
		if _, err := io.ReadFull(client, pBuf); err != nil {
			return
		}
		if string(uBuf) != user || string(pBuf) != pass {
			client.Write([]byte{0x01, 0xFF}) // auth failure
			return
		}
		client.Write([]byte{0x01, 0x00}) // auth success
	} else {
		// No auth required.
		client.Write([]byte{0x05, 0x00})
	}

	// Request
	req := make([]byte, 4)
	if _, err := io.ReadFull(client, req); err != nil {
		return
	}
	if req[0] != 0x05 || req[1] != 0x01 { // only CONNECT
		client.Write([]byte{0x05, 0x07, 0x00, 0x01, 0, 0, 0, 0, 0, 0})
		return
	}

	var host string
	switch req[3] {
	case 0x01: // IPv4
		addr := make([]byte, 4)
		io.ReadFull(client, addr)
		host = net.IP(addr).String()
	case 0x03: // Domain
		lenBuf := make([]byte, 1)
		io.ReadFull(client, lenBuf)
		domain := make([]byte, lenBuf[0])
		io.ReadFull(client, domain)
		host = string(domain)
	case 0x04: // IPv6
		addr := make([]byte, 16)
		io.ReadFull(client, addr)
		host = net.IP(addr).String()
	default:
		client.Write([]byte{0x05, 0x08, 0x00, 0x01, 0, 0, 0, 0, 0, 0})
		return
	}

	portBuf := make([]byte, 2)
	if _, err := io.ReadFull(client, portBuf); err != nil {
		return
	}
	port := binary.BigEndian.Uint16(portBuf)
	target := fmt.Sprintf("%s:%d", host, port)

	// Connect to target
	remote, err := net.Dial("tcp", target)
	if err != nil {
		client.Write([]byte{0x05, 0x04, 0x00, 0x01, 0, 0, 0, 0, 0, 0})
		return
	}
	defer remote.Close()

	// Success response
	localAddr := remote.LocalAddr().(*net.TCPAddr)
	resp := []byte{0x05, 0x00, 0x00, 0x01}
	resp = append(resp, localAddr.IP.To4()...)
	resp = append(resp, byte(localAddr.Port>>8), byte(localAddr.Port))
	client.Write(resp)

	// Bidirectional relay
	done := make(chan struct{}, 2)
	go func() { io.Copy(remote, client); done <- struct{}{} }()
	go func() { io.Copy(client, remote); done <- struct{}{} }()
	<-done
}

File: agents/agent-rust/src/socks.rs
#![allow(dead_code)]
/// Minimal SOCKS5 proxy — no-auth only, CONNECT command only.
/// socks_start(port) binds the listener and accepts in a background thread.
/// socks_stop() signals the accept loop to exit.

use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream};
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Duration;

static SOCKS_STOP: AtomicBool = AtomicBool::new(false);

// ── SOCKS5 client handler ─────────────────────────────────────────────────────

fn handle_socks5(mut client: TcpStream) {
    client.set_nonblocking(false).ok();

    // ── 1. Auth negotiation (RFC 1928 §3) ────────────────────────────────────
    let mut hdr = [0u8; 2];
    if client.read_exact(&mut hdr).is_err() { return; }
    if hdr[0] != 0x05 { return; } // must be SOCKS5
    let nmethods = hdr[1] as usize;
    let mut methods = vec![0u8; nmethods];
    if nmethods > 0 && client.read_exact(&mut methods).is_err() { return; }
    // Always reply: version=5, method=0 (no auth)
    if client.write_all(&[0x05, 0x00]).is_err() { return; }

    // ── 2. CONNECT request (RFC 1928 §4) ────────────────────────────────────
    let mut req = [0u8; 4];
    if client.read_exact(&mut req).is_err() { return; }
    if req[0] != 0x05 || req[1] != 0x01 { return; } // only CONNECT(1)

    let target_host: String = match req[3] {
        0x01 => {
            // IPv4
            let mut addr = [0u8; 4];
            if client.read_exact(&mut addr).is_err() { return; }
            std::net::Ipv4Addr::from(addr).to_string()
        }
        0x03 => {
            // Domain name
            let mut lenbuf = [0u8; 1];
            if client.read_exact(&mut lenbuf).is_err() { return; }
            let mut domain = vec![0u8; lenbuf[0] as usize];
            if client.read_exact(&mut domain).is_err() { return; }
            String::from_utf8_lossy(&domain).into_owned()
        }
        0x04 => {
            // IPv6
            let mut addr = [0u8; 16];
            if client.read_exact(&mut addr).is_err() { return; }
            std::net::Ipv6Addr::from(addr).to_string()
        }
        _ => return,
    };
    let mut port_buf = [0u8; 2];
    if client.read_exact(&mut port_buf).is_err() { return; }
    let port = u16::from_be_bytes(port_buf);

    let target = format!("{}:{}", target_host, port);

    // ── 3. Open upstream connection ──────────────────────────────────────────
    let server = match TcpStream::connect(&target) {
        Ok(s) => s,
        Err(_) => {
            // Connection refused reply
            let _ = client.write_all(&[0x05, 0x05, 0x00, 0x01, 0, 0, 0, 0, 0, 0]);
            return;
        }
    };

    // Success reply: VER=5 REP=0 RSV=0 ATYP=1 BND.ADDR=0.0.0.0 BND.PORT=0
    if client.write_all(&[0x05, 0x00, 0x00, 0x01, 0, 0, 0, 0, 0, 0]).is_err() { return; }

    // ── 4. Bidirectional relay ───────────────────────────────────────────────
    let client_for_thread = match client.try_clone() { Ok(c) => c, Err(_) => return };
    let server_for_main   = match server.try_clone() { Ok(s) => s, Err(_) => return };

    // Thread: client → server
    let t = std::thread::spawn(move || {
        let mut r = client_for_thread;
        let mut w = server;
        let _ = std::io::copy(&mut r, &mut w);
        let _ = w.shutdown(std::net::Shutdown::Both);
    });

    // Current thread: server → client
    let mut r = server_for_main;
    let mut w = client;
    let _ = std::io::copy(&mut r, &mut w);
    let _ = w.shutdown(std::net::Shutdown::Both);
    let _ = t.join();
}

// ── Public API ────────────────────────────────────────────────────────────────

pub fn socks_start(port: u16) -> Result<(), String> {
    SOCKS_STOP.store(false, Ordering::Relaxed);

    let listener = TcpListener::bind(("0.0.0.0", port))
        .map_err(|e| format!("socks bind: {}", e))?;
    listener.set_nonblocking(true)
        .map_err(|e| format!("set_nonblocking: {}", e))?;

    std::thread::spawn(move || {
        loop {
            if SOCKS_STOP.load(Ordering::Relaxed) { break; }
            match listener.accept() {
                Ok((client, _)) => {
                    std::thread::spawn(move || handle_socks5(client));
                }
                Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                    std::thread::sleep(Duration::from_millis(100));
                }
                Err(_) => break,
            }
        }
    });

    Ok(())
}

pub fn socks_stop() {
    SOCKS_STOP.store(true, Ordering::Relaxed);
}
